src/rustc/rustc.rs

```diff
@@ -11,17 +10,32 @@
 // Also, don't forget to set this for rustdoc.
 extern {}
 
-extern crate rustc_driver;
-
-// Note that the linkage here should be all that we need, on Linux we're not
-// prefixing the symbols here so this should naturally override our default
-// allocator. On OSX it should override via the zone allocator. We shouldn't
-// enable this by default on other platforms, so other platforms aren't handled
-// here yet.
-#[cfg(feature = "jemalloc-sys")]
-extern crate jemalloc_sys;
-
 fn main() {
+    // Pull in jemalloc when enabled.
+    //
+    // Note that we're pulling in a static copy of jemalloc which means that to
+    // pull it in we need to actually reference its symbols for it to get
+    // linked. The two crates we link to here, std and rustc_driver, are both
+    // dynamic libraries. That means to pull in jemalloc we need to actually
+    // reference allocation symbols one way or another (as this file is the only
+    // object code in the rustc executable).
+    //
+    // As such, here's an interprative dance to slip under LLVM's radar yet
+    // please the linker.
+    #[cfg(feature = "jemalloc-sys")] {
+        if std::env::var("__RUSTC_NEVER_PRESENT_ENV_VAR").is_ok() {
+            unsafe {
+                jemalloc_sys::calloc(1, 4);
+                jemalloc_sys::posix_memalign(&mut std::ptr::null_mut(), 1, 10);
+                jemalloc_sys::aligned_alloc(1, 10);
```

Comment:
I confess this goes over my head. I guess that my keeping these calls live, we keep things linked in, and then calls to `calloc` and so forth wind up invoking jemalloc and not the system allocator?

Reply:
Right yeah, we need to give the linker a reason to actually statically link jemalloc, which means we need to reference allocation symbols which will get satisfied by jemalloc's library.

Reply:
@alexcrichton 
Hmm, this reminded me of one question I wanted to ask.
Static libraries linked through command line `-l static=mylib` are wrapped into `--whole-archive` (which would obviate the need to add these dummy uses for otherwise unused symbols).
Static libraries linked from source with `#[link(name = "test", kind = "static")]`, on the other hand, are not wrapped into `--whole-archive`.
What's the history behind this?

Reply:
Hm, actually `jemalloc-sys` links jemalloc though command line, then why are the symbols lost?
Do I understand correctly that static `jemalloc` is linked into `jemalloc-sys.rlib` (?) (it uses `static`, rather than `static-nobundle`) and the symbols are not lost at *that* moment.
The symbols are lost later than static `jemalloc-sys` is linked into rustc executable, and `--whole-archive` is not used at this time?

Reply:
I think the `--whole-archive` with `-l` vs `#[link]` is probably a bug actually? I wasn't aware of that behavior. Unfortunately though it wouldn't help in this case because jemalloc is included in `libjemalloc_sys.rlib` which isn't linked with `--whole-archive`

(we don't link rlibs with `--whole-archive`)

>  Do I understand correctly that static jemalloc is merged into jemalloc-sys.rlib (it uses static, rather than static-nobundle) and the symbols are not lost at that moment.

correct

>  The symbols are lost later when static jemalloc-sys is linked into rustc executable, and --whole-archive is not used at this time?

correct!